### PR TITLE
Add code references

### PR DESCRIPTION
**Pull request by namanjain**
This is the next step towards hashing code references. The
implementation is copied from Streamlit, which is a reasonably popular
framework. But I added some basic tests to be sure that it works
correctly.

It also integrates the code references logic in Bionic's caching
strategy. It gets us a step closer to having a robust automatic caching
strategy.

Here is the list of things left after this change:

- hashing classes
- dedup references
- maybe return references that are stored in local variable
- caching individual object hashes for a CodeHasher run
- investigate if we can hash module or package versions
- verify that we hash all Python constant types
- version.suppress_bytecode_warnings TODO
- skip referenced code objects
- Only call getattr for modules TODO
- Throw an exception for unhandled bytecode instruction type

### CODE REVIEW

File: bionic/code_references.py
Comment (namanjain) [inline bionic/code_references.py:122]: I wasn't sure if I should throw an error or warning here, but I ended up using a warning. I think it's really bad if a user can't use the flow at all. Maybe we should give more information like how to delete / invalidate the cache entries manually when they see this warning. That way, the users won't be completely blocked. Wdyt? 

FWIW, Streamlit throws an error here.

Comment (jqmp) [reply bionic/code_references.py:122]: It sounds like we're going to keep iterating through the remaining opcodes. Does that make sense? Will our state still be valid?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:63]: Do we have a test case that illustrates this `foo.bar` case and how it's useful? I see the example with `[MyClass, "log_val"]`, but it's not clear to me why we care about `"log_val"`, since presumably we can get that just from hashing the bytecode.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:122]: Now that I had some more time to think about it, I think I should change the catch block and add the tos to references if it contains one.

File: bionic/code_references.py
Comment (jqmp) [reply bionic/code_references.py:47]: 1. What is the definition of "major"? And why?
2. If we don't catch every kind of reference, why not?
3. How do we know that it will catch every major reference? I'd like to be more confident than "I tried as many tests as I could think of and they passed". We should be able to read the code and be convinced that it is correct. What can you tell the reader that would convince them?

Comment (namanjain) [reply bionic/code_references.py:74]: I added docstring to this method. But I think it's really hard to justify `3.` without explaining how python bytecode instructions work.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py]: After thinking about it a bit more (and talking on slack), I changed it to an error (like Streamlit).

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:157]: Like we discussed a while back, I think we should give users the option to ignore bytecode of an entity. Since we don't have that yet, I haven't added anything related to it in the error message. But I can add a code comment (or a TODO) so we don't forget about it.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:126]: I think this is just defensive coding. I'll remove it from our implementation.

File: tests/test_code_references.py
Comment (namanjain) [reply tests/test_code_references.py:1]: Streamlit doesn't have any unit tests for finding code references, they only have integration tests. But looking at their [integration tests](https://github.com/streamlit/streamlit/blob/develop/lib/tests/streamlit/hashing_test.py#L649), I think I'm covering all of the cases they do.

I'll add all the examples in the appendix of the Code References doc here to have more complex cases.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:74]: I added some more details in the function docstring.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:60]: Streamlit never explained it anywhere, but I left my understanding of this as a code comment.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:44]: No... I don't know how that ended up there...

File: tests/test_code_references.py
Comment (namanjain) [reply tests/test_code_references.py:186]: Yup, that doesn't matter. The purpose of the inner function was to test for cell variable. The `inner_x` function is also a constant for `enclosing_x` function which will be hashed separately, so any references inside `inner_x` will be taken care of then.

I think this would become apparent when we integrate code referencing logic with our hashing logic and tests for them.

File: tests/test_code_references.py
Comment (namanjain) [reply tests/test_code_references.py:174]: We will go through bytecode instructions for every line of every control flow. I'll add a test for loops and if else. 👍🏽

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:151]: For posterity, we discussed this off-GitHub. Importing a module like this only happens when someone uses import statements inside functions. If a user is importing modules inside a function, it's mostly because they don't want to import the module till function execution time.

In Bionic, we will change this and not execute import statements either. Instead, we will suggest users to make the import statement global if possible. We will also emit a warning and give users an option to either suppress the warning or turn off bytecode analysis altogether.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:161]: I don't think it's possible for attribute to not exist. Python should throw an error during compile time in that case.

But yes, if the attribute is a property, it will run the property method code. I didn't find a way to detect whether a variable is a property.

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:147]: Thinking about this more: here `op.argval` is just a string, right? So if later we have a `LOAD_ATTR` instruction, we will call `getattr` on this string? Is that what we want?

Comment (jqmp) [reply bionic/code_references.py:147]: Also, it might be helpful to have a comment explaining how this case arises. If we see a `LOAD_NAME` and it's not in the global names, what other places could it be in?

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:161]: I was very off mark about python compile time checks. For now, I'm going to gate this behind `hasattr` check.

We also discussed this off-GitHub but we might end up calling `getattr` only for modules. But that would require some more thought to make sure we don't miss anything. I'll leave that as a TODO.

Comment (namanjain) [reply bionic/code_references.py:161]: I'm gonna leave a TODO for properties too.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:147]: If we get `LOAD_ATTR` when `TOS` is string, we concatenate the TOS with the attribute name.

From my understanding of Python and bytecode, I think this can only happen if the variable does not exist. If you look at the test `test_function_references`, you'll see that the last example in that test return a `"func_does_not_exist"` string because the function does not exist and cannot be found.

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:109]: I don't think "Smart Caching" is referenced anywhere else in our codebase. I don't think you need to go into a lot of detail here, but can we write this in a more standalone way?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:151]: nit: s/it's/its/

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:150]: What does it mean for the variable to not exist? Are you saying this can only happen if the code is referencing an invalid name? It looks to me like `LOAD_NAME` is intended to look up more than just globals, so presumably this could be a valid reference even if it's not in `context.globals`?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:165]: What if they can't import it globally for some reason? What should they do to avoid this warning?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:179]: It sounds like we intend to complete this TODO before releasing this code -- is that right? If so, can we document this issue in a more prominent way to make sure we remember to fix it?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:188]: I hadn't noticed before that we're mutating the context object here. Should we make this clear in the docstring? Would we ever continue using the context object after calling this function?

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:212]: Thinking about this more: should this be a `CodeVersioningError`? If this a bug in Bionic, maybe it should be an `AssertionError`?

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:149]: You're right, it looks `LOAD_NAME` also looks at the local variables in the frame object (which we don't have). I'll update this.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:205]: We discussed this sometime back (before I started working on the integration logic). I think we decided to not delete variables, so I'm going to remove `DELETE_FAST` branch from here.

Comment (namanjain) [reply bionic/code_references.py:205]: I think we should `DELETE_FAST` anyway, but just to be sure that there is no side effect of mutating the context object during this analysis phase, I'll just make a copy of `code_context` at the beginning of this method.

File: tests/test_flow/test_persistence.py
Comment (jqmp) [inline tests/test_flow/test_persistence.py:845]: What is `warns(None)` doing? My understanding is that this just records the warnings for later use, but we're not using them, are we?

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:55]: I don't think there is anything useful we can do with the context when it is false. I was leaning towards keeping it None and asserting the value but this looked a bit cleaner when I tried. I can move back to None and assert though. Wdyt?

File: tests/test_flow/test_persistence.py
Comment (namanjain) [reply tests/test_flow/test_persistence.py:911]: Ohhh, I thought it asserts that the warning is `None`. But you are right, it records the warnings. I'll check that there are no recorded warnings with `pytest.warns(None)`.

File: tests/test_code_references.py
Comment (jqmp) [inline tests/test_code_references.py:212]: If this test finds a bug in `CodeHasher`, then fixing the bug will also change the test, right?

Comment (namanjain) [reply tests/test_code_references.py:212]: Changing `CodeHasher` code will likely change the returned references in this test too. Even if we don't add any new references, we will probably end up with a different order of the references. 

I tried a few methods from our test code as well (like the random tests), but it was referencing internal pytest methods and it was easier to use a complex method from Bionic code instead. I can always use a different method (replacing this or in addition to this) if you have something in mind.

Instead of just copying the returned references, I wrote all the references myself and then checked that the returned references were the same. This way I was able to make sure that we didn't miss a reference. I did add a TODO in this test which points that we might miss references from a conditional block statement. That's on my top of list of things to follow up on.

Comment (namanjain) [reply tests/test_code_references.py:212]: Oh and if your concern is that the test is testing it's own source code, you shouldn't be worried about that. We are testing the logic for `get_code_references` by calling it on `CodeHasher`, which is separate from `get_code_references`.

File: bionic/code_references.py
Comment (jqmp) [reply bionic/code_references.py:55]: It seems to me that there are two separate issues:

1. If we don't have a valid code context, should we represent that with `None` or with a special flag on the context object? I think `None` is preferable to having a flag that means "don't use this object for anything".
2. If we don't have a valid code context when processing a code object, should we throw an exception or just skip analyzing the code? I'm not sure what the right answer is. What do you think?

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:98]: Sorry, forgot to reply to here. The example in the docstring will only return two references: 
* the function get_my_class, and
* the string `"call"`

The string `"MyClass"` will be a constant and hashed by `CodeHasher` from `func.__code__.co_consts` variable.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:216]: I'm keeping track of things left to do separately. This TODO is just a reminder in case it slips through the cracks (which it shouldn't).

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:55]: 1. I was leaning towards `None` too but this seemed a bit cleaner in case we want to ever change it later. I can go back to `None`.

2. I'm on the fence for this one too. I went from throwing an exception to ignoring it, but we won't ever get to this point unless someone passes a code object to the hasher directly (which shouldn't happen) or defines a code object themselves (do people even do this?). I can be easily convinced otherwise, it is totally reasonable to throw an exception as a guard against any future bugs. But if you don't feel strongly, I think I'll keep ignoring it in case anyone ever defines a code object by themself.

File: tests/test_code_references.py
Comment (jqmp) [reply tests/test_code_references.py:212]: Oh right, I was wrong to think that changing CodeHasher would affect this test.

That said, this does seem to have the potential to introduce a lot of annoying busywork whenever we change the code of `CodeHasher`. I don't think this needs to block this PR though.

File: bionic/datatypes.py
Comment (jqmp) [inline bionic/datatypes.py:304]: I would phrase the last sentence as, "If `include_bytecode` is True, then the major version is understood to implicitly include the bytecode of the code as well."

File: docs/concepts.rst
Comment (jqmp) [inline docs/concepts.rst:325]: I guess this should say "Built-In or Installed Modules". Also, "built-in modules or installed modules" should probably be "built-in or installed modules or functions".

File: docs/concepts.rst
Comment (jqmp) [inline docs/concepts.rst:327]: We're missing a close parenthesis here.

File: bionic/code_references.py
Comment (jqmp) [inline bionic/code_references.py:93]: Just realized this: doesn't returning strings as placeholders also introduce the possibility of collisions? Like if we change this code

    x = call_func() # Suppose we can't following this reference

to 

    x = "call_func"

Can a situation like this produce the same hash?

WDYT about wrapping these placeholder strings in a special class? This might also solve the issue of using the `tos` variable for both reference values and reference names.

**Review by jqmp — COMMENTED** [review]
I took another pass through the code. Most of the remaining changes seem pretty small. However, it seems like there are still several followup changes we plan to make that will change the hashes, which means that if we make those changes after releasing this code then we'll have another cache-breaking release. There are also some unaddressed edge cases that I'm not sure we'd want to release without fixing.

Normally I think it would be good practice to just fix all these things before merging, but I know this PR has been outstanding for a while. Could you just add a list of all the stuff that should be fixed before release and add it as a comment or something?

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:49]: We don't need these now. I'll revert the changes.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:102]: Right, it should be "and". Good point about local, I'll remove that.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:93]: I don't think an example like that will result in possible collisions because we also hash the `func.__code__.co_code` variable which is the bytecode hash of the function itself. Those two functions will have different `co_code`.

We can hash those strings in a separate class, but we will have to wait till we hash classes for it to work. I'll leave a TODO to change it once we have support for hashing classes.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:220]: I think this is because it stores the `tos` in `context.varnames`. But I agree that maybe we should use `set_tos(None)` and return the current `tos` as a reference. This is exactly what I meant to do from this [TODO](https://github.com/square/bionic/pull/232/files#diff-9e61581018162d3520e8e837df29a47bb3bea1afc6e11643b25b66d4e67d4a03R254).

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:160]: Yeah, we can have a list of known instructions at the end and throw an error if we encounter an unknown instruction. This is becoming a theme now, but I'll leave another TODO.

File: bionic/code_references.py
Comment (namanjain) [reply bionic/code_references.py:204]: We didn't and it's a good point. But like you said, your suggestion in [another comment](https://github.com/square/bionic/pull/232#discussion_r542581595) should fix the issue.

File: docs/concepts.rst
Comment (jqmp) [inline docs/concepts.rst:353]: nit: I would say "*these* false negatives" to connect back to the previous section.

**Review by jqmp — APPROVED** [review]

**Pull request merged**
